### PR TITLE
coord: correctly handle TYPE dependencies

### LINKED ISSUE #5577: DROP TYPE allowed even if objects still reference said type -> Materialize fails to load catalog on startup
```
pstoev=> CREATE TYPE int4_list AS LIST (element_type = int4);
CREATE TYPE

pstoev=> create view v1 as select cast('{2}' as int4_list);
CREATE VIEW

pstoev=> drop type int4_list;
DROP TYPE
```

Restart than fails:

```
materialized: corrupt catalog: failed to deserialize item u2153 (pstoev.public.v1): unknown catalog item 'pstoev.public.int4_list'
```

Rather than tracking dependencies exhaustively, this can be resolved by in-lining the complete type definition inside every view that references it. This will however prevent a proper future implementation of e.g. ```ALTER TYPE```

### PR DESCRIPTION
**Pull request by JLDLaughlin**
Fixes #5577.

**To do**: tests for the other `CatalogItems` that might depend on custom types: `Source`, `Sink`, and `Func`.

---
This change is [Reviewable](https://reviewable.io/reviews/materializeinc/materialize/5679)

### CODE REVIEW

**Comment by benesch** [comment]
@sploiselle you mind taking this review when @JLDLaughlin says it's ready?

File: src/coord/src/catalog.rs [hunk 1]
```diff
@@ -1013,6 +1013,48 @@ impl Catalog {
...
 
+    pub fn custom_oid_dependencies(&self, item: &CatalogItem) -> Option<Vec<u32>> {
+        match item {
+            CatalogItem::Table(Table { desc, .. }) | CatalogItem::Source(Source { desc, .. }) => {
+                Some(desc.typ().custom_type_oids())
+            }
+            CatalogItem::View(View { optimized_expr, .. }) => {
+                Some(optimized_expr.clone().into_inner().typ().custom_type_oids())
+            }
+            CatalogItem::Index(Index { on, keys, .. }) => {
```
Comment (philip-stoev) [inline src/coord/src/catalog.rs:1024]: For some reason, ```DROP TYPE``` appears to drop the entire index where it is used(?):

```
CREATE TYPE int4_list AS LIST (element_type=int4);
create table t1 (f1 int4_list);
 CREATE TYPE int4_list2 AS LIST (element_type=int4);
create index i1 on t1 ( CAST(f1 AS int4_list2));
materialize=> show index from t1;
 on_name |    key_name    | seq_in_index | column_name |            expression             | nullable 
---------+----------------+--------------+-------------+-----------------------------------+----------
 t1      | i1             |            1 |             | f1::materialize.public.int4_list2 | f
 t1      | t1_primary_idx |            1 | f1          |                                   | t
(2 rows)

materialize=> drop type int4_list2; <- this should have failed
DROP TYPE
materialize=> show index from t1;
 on_name |    key_name    | seq_in_index | column_name | expression | nullable 
---------+----------------+--------------+-------------+------------+----------
 t1      | t1_primary_idx |            1 | f1          |            | t
```

File: src/coord/src/catalog.rs [hunk 2]
```diff
@@ -1013,6 +1013,48 @@ impl Catalog {
         }
     }
 
+    pub fn custom_oid_dependencies(&self, item: &CatalogItem) -> Option<Vec<u32>> {
+        match item {
+            CatalogItem::Table(Table { desc, .. }) | CatalogItem::Source(Source { desc, .. }) => {
+                Some(desc.typ().custom_type_oids())
+            }
+            CatalogItem::View(View { optimized_expr, .. }) => {
```
Comment (philip-stoev) [inline src/coord/src/catalog.rs:1021]: I am afraid the check does not recurse into the expressions correctly. This ```DROP TYPE``` is allowed to succeed and borks the catalog:

```
materialize=> CREATE TYPE int4_list AS LIST (element_type=int4)
materialize-> ;
CREATE TYPE
materialize=> CREATE VIEW v1 AS SELECT CAST(CAST('{2}' AS int4_list) AS text); 
CREATE VIEW
materialize=> drop type int4_list;
DROP TYPE
```

This one too:

```
materialize=> CREATE VIEW v1 AS SELECT MIN(CAST(CAST('{1}' AS int4_list) AS string));
CREATE VIEW
materialize=> drop type int4_list;
DROP TYPE
```

File: src/coord/src/catalog.rs [hunk 3]
```diff
@@ -1013,6 +1013,48 @@ impl Catalog {
         }
     }
 
+    pub fn custom_oid_dependencies(&self, item: &CatalogItem) -> Option<Vec<u32>> {
+        match item {
+            CatalogItem::Table(Table { desc, .. }) | CatalogItem::Source(Source { desc, .. }) => {
```
Comment (philip-stoev) [inline src/coord/src/catalog.rs:1018]: The ```DEFAULT``` clause for tables is not being checked:

```materialize=> CREATE TYPE int4_list AS LIST (element_type=int4)
materialize-> ;
CREATE TYPE
materialize=>  CREATE TABLE t1 (f1 TEXT DEFAULT CAST ('{}' AS int4_list));
CREATE TABLE
materialize=> DROP TYPE int4_list;
DROP TYPE
```

**Comment by philip-stoev** [comment]
I pushed as many tests as I could think of and some will currently fail. As I was unable to run the .td file to completion, I wrote them blindly, so there may be errors or typos in the tests themselves.

I could not come up with any tests for Source, Sink, and Func though. I do not see any type information being passed in Source or Sink definitions and my understanding is the SQL scalar functions are still to be implemented?

**Comment by JLDLaughlin** [comment]
thanks @philip-stoev ! I'll work through these today

**Comment by JLDLaughlin** [comment]
Closing in favor of https://github.com/MaterializeInc/materialize/pull/5810

**Pull request closed without merging**